pkg/location/extensions/extensions.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extensions
+
+import (
+	"context"
+	"fmt"
+)
+
+type Reference interface {
+	fmt.Stringer
+	Type() string
+	Validate() error
+}
+
+type ReferenceLock interface {
+	Reference
+}
+
+type IdentifierGetter interface {
+	GetIdentifier() (string, bool)
+}
+
+type LockGetter interface {
+	GetLock() (string, bool)
+}
+
+// DefaultDirectoryNameGetter is present on Reference types that
+// suggest a default local folder name
+type DefaultDirectoryNameGetter interface {
+	// GetDefaultDirectoryName implements the location.DefaultDirectoryName() method
+	GetDefaultDirectoryName() (string, bool)
+}
+
+// DefaultIdentifierGetter is present on Reference types that
+// suggest a default Identifier.
+type DefaultIdentifierGetter interface {
+	GetDefaultIdentifier(ctx context.Context) (string, error)
+}
+
+// RelPather is present on Reference types that
+// will return a relative path if one reference is a sub-package
+// location in another. The comparison is strict, meaning all criteria
+// other than the directory component (like repo, ref, image, tag, etc.) must be equal.
+type RelPather interface {
+	Rel(targref Reference) (string, error)
```

Comment:
shouldn't it be returning `location.Reference` instead of string ?

Reply:
In this case the difference between two Reference would be a relative path - which is the string return value. The place it's used is in `kpt pkg update` in the "should update sub package" where the logic is asking if the nested kptfile.upstream is pointing to exactly the same remote location - but in a folder that's nested.

See also pkg/location/reference.go `func Rel(baseref Reference, targref Reference) (string, error)` which is the api that's actually called (not the RelPather directly)

Actually a `type RelativePath string` could be nice to have, but that's also a huge change. Maybe a simpler update would be to have named return arguments like this?

func Rel(baseref Reference, targref Reference) (relPath string, err error)` 

The argument names baseref and targref are not great - they came straight from filepath.Rel argument names. Suggestions welcomed there as well.